torch/csrc/utils/tensor_types.cpp

```diff
@@ -83,11 +83,18 @@ at::TensorOptions options_from_string(const std::string& str) {
 
 std::vector<std::pair<Backend, ScalarType>> all_declared_types() {
   std::vector<std::pair<Backend, ScalarType>> ret;
-  // can't easily iterate over enum classes
+
+  // Can't easily iterate over enum classes.
   std::vector<Backend> backends = { Backend::CPU, Backend::CUDA, Backend::SparseCPU, Backend::SparseCUDA };
-  std::vector<ScalarType> scalar_types = { ScalarType::Byte, ScalarType::Char, ScalarType::Double, ScalarType::Float,
-                                           ScalarType::Int, ScalarType::Long, ScalarType::Short, ScalarType::Half,
-                                           ScalarType::Bool, ScalarType::BFloat16};
+  // Try to keep in sync with ScalarType.  Note how this doesn't use
+  // AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_AND_QINTS as ScalarType does because
+  // this list is used for initializing empty tensors, and that doesn't work
+  // with qint.
+  std::vector<ScalarType> scalar_types = {
+    #define DEFINE_ENUM(_1, n) ScalarType::n,
+      AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF(DEFINE_ENUM)
+  };
```

Comment:
I think `torch.ComplexDoubleTensor`, `torch.cuda.ComplexDoubleTensor` etc. intentionally wasn't added because those types are legacy. Instead, I think `py_set_default_dtype` should avoid the intermediate step through the legacy tensor type api.

Reply:
ouch, i pushed a new set of changes before seeing this comment

@anjali411 wdyt? do we rollback to the previous set of types and just add more asserts in the function that failed initially? either way, the thing that we're doing rn is error-prone because it's not actually all-types (the macros are different). so if more types are added later, it'll fall apart again :/

Reply:
@anjali411 just to be 100% clear. This PR has the non-obvious effect of adding 4 new classes to the public python API: `torch.ComplexFloatTensor`, `torch.ComplexDoubleTensor`, `torch.cuda.ComplexFloatTensor` and `torch.cuda.ComplexDoubleTensor`.

I'm not sure this was the desired outcome, so want to make sure.

Reply:
cc @albanD